Barotrauma/BarotraumaShared/SharedSource/LuaCs/Plugins/AssemblyManager.cs

```diff
@@ -0,0 +1,629 @@
+    {
+        get
+        {
+            OpsLockUnloaded.EnterReadLock();
+            try
+            {
+                return UnloadingACLs.Any();
+            }
+            catch (Exception)
+            {
+                return false;
+            }
+            finally
+            {
+                OpsLockUnloaded.ExitReadLock();
+            }
+        }
+    }
+    
+    
+    /// <summary>
+    /// Allows iteration over all non-interface types in all loaded assemblies in the AsmMgr that are assignable to the given type (IsAssignableFrom).
+    /// </summary>
+    /// <typeparam name="T">The type to compare against</typeparam>
+    /// <returns>An Enumerator for matching types.</returns>
+    public IEnumerable<Type> GetSubTypesInLoadedAssemblies<T>()
+    {
+        Type targetType = typeof(T);
+
+        OpsLockLoaded.EnterReadLock();
+        try
+        {
+            return AssemblyLoadContext.Default.Assemblies
+                .SelectMany(a => a.GetSafeTypes())
+                .Where(t => targetType.IsAssignableFrom(t) && !t.IsInterface)
+                .Concat(LoadedACLs
+                    .SelectMany(kvp => kvp.Value.GetAssembliesTypes()
+                        .Where(t => targetType.IsAssignableFrom(t) && !t.IsInterface)))
+                .ToImmutableList();
+        }
+        finally
+        {
+            OpsLockLoaded.ExitReadLock();
+        }
+    }
+
+    /// <summary>
+    /// Tries to get types assignable to type from the ACL given the Guid.
+    /// </summary>
+    /// <param name="id"></param>
+    /// <param name="types"></param>
+    /// <typeparam name="T"></typeparam>
+    /// <returns></returns>
+    public bool TryGetSubTypesFromACL<T>(Guid id, out IEnumerable<Type> types)
+    {
+        Type targetType = typeof(T);
+
+        if (TryGetACL(id, out var acl))
+        {
+            types = acl.GetAssembliesTypes()
+                .Where(t => targetType.IsAssignableFrom(t) && !t.IsInterface);
+            return true;
+        }
+
+        types = null;
+        return false;
+    }
+    
+    /// <summary>
+    /// Tries to get types from the ACL given the Guid.
+    /// </summary>
+    /// <param name="id"></param>
+    /// <param name="types"></param>
+    /// <typeparam name="T"></typeparam>
+    /// <returns></returns>
+    public bool TryGetSubTypesFromACL(Guid id, out IEnumerable<Type> types)
+    {
+        if (TryGetACL(id, out var acl))
+        {
+            types = acl.GetAssembliesTypes();
+            return true;
+        }
+
+        types = null;
+        return false;
+    }
+
+
+    /// <summary>
+    /// Allows iteration over all non-interface types in all loaded assemblies in the AsmMgr who's names contain the string.
+    /// </summary>
+    /// <param name="name">The string name of the type to search for, search uses EndsWith()</param>
+    /// <returns>An Enumerator for matching types.</returns>
+    public IEnumerable<Type> GetTypesByName(string name)
```

Comment:
I agree with this but see below for why this change was put into Modding Toolkit in the first place.


There was an issue where Lua users were complaining that their custom ItemComponents were not being found in loaded assemblies because they were *always* searching using partial name.

Part of this reason was because of the LuaCsForBarotrauma helper function given out was prepending "Barotrauma." in the helper function Evil provided to every type search string, meaning that all custom Types had to be declared in the Barotrauma namespace, which is less than ideal and not what should be encouraged (I ran into this the most while working in Mechtrauma dev).

Even after I provided a search function for custom assemblies and Barotrauma only, they still had issues because "Barotrauma." was no longer prepended but adding it back leads to the previous problem.

The interim solution was to use `.EndsWith()` which allowed "fuzzy" searches for a type and worked 99% of the time. If necessary the fully qualified name can still be used. The idea was that if a user ran into an ambiguous case, they could just specify the full name of the type and get intended type.

Reply:
Fixed.